services/client/services.go

```diff
@@ -96,129 +193,257 @@ func (s *client) Query(sid services.ServiceID, opts services.QueryOptions) (serv
+		c.opts.InstrumentsOptions().MetricsScope().Tagged(map[string]string{
 			"sd_service": sid.Name(),
 			"sd_env":     sid.Environment(),
 			"sd_zone":    sid.Zone(),
 		}),
 	)
-	go s.run(valueWatch, watchable, sid, sdm.serviceUnmalshalErr)
-	go updateVersionGauge(valueWatch, sdm.versionGauge)
+	go c.run(placementWatch, watchable, sid, opts, hbStore, initService, ids, sdm.serviceUnmalshalErr)
+	go updateVersionGauge(placementWatch, sdm.versionGauge)
 
 	_, w, err := watchable.Watch()
 	return w, err
 }
 
-func (s *client) placement(sid services.ServiceID) (kv.Value, error) {
-	pMgr, err := s.getPlacementManager(sid.Zone())
+func (c *client) getPlacementValue(sid services.ServiceID) (kv.Value, error) {
+	kvm, err := c.getKVManager(sid.Zone())
 	if err != nil {
 		return nil, err
 	}
 
-	v, err := pMgr.kv.Get(placementKey(sid.Environment(), sid.Name()))
+	v, err := kvm.kv.Get(placementKey(sid))
 	if err != nil {
 		return nil, err
 	}
 
 	return v, nil
 }
 
-func (s *client) getPlacementManager(zone string) (*placementManager, error) {
-	s.Lock()
-	defer s.Unlock()
-	pManager, ok := s.pManagers[zone]
+func (c *client) getHeartbeatStore(zone string) (heartbeat.Store, error) {
+	c.Lock()
+	defer c.Unlock()
+	hb, ok := c.hbStores[zone]
+	if ok {
+		return hb, nil
+	}
+
+	hb, err := c.opts.HBGen()(zone)
+	if err != nil {
+		return nil, err
+	}
+
+	c.hbStores[zone] = hb
+	return hb, nil
+}
+
+func (c *client) getKVManager(zone string) (*kvManager, error) {
+	c.Lock()
+	defer c.Unlock()
+	m, ok := c.kvManagers[zone]
 	if ok {
-		return pManager, nil
+		return m, nil
 	}
 
-	kv, err := s.opts.KVGen()(zone)
+	kv, err := c.opts.KVGen()(zone)
 	if err != nil {
 		return nil, err
 	}
 
-	pManager = &placementManager{
+	m = &kvManager{
 		kv:                kv,
 		serviceWatchables: map[string]xwatch.Watchable{},
 	}
 
-	s.pManagers[zone] = pManager
-	return pManager, nil
+	c.kvManagers[zone] = m
+	return m, nil
 }
 
-func (s *client) run(vw kv.ValueWatch, w xwatch.Watchable, sid services.ServiceID, errCounter tally.Counter) {
-	for range vw.C() {
-		value := vw.Get()
-		s.opts.InstrumentsOptions().Logger().Infof("received topology update notification on version %d", value.Version())
+func (c *client) run(
+	vw kv.ValueWatch,
+	w xwatch.Watchable,
+	sid services.ServiceID,
+	qopts services.QueryOptions,
+	hbStore heartbeat.Store,
+	service services.Service,
+	ids []string,
+	errCounter tally.Counter,
+) {
+	ticker := time.NewTicker(c.opts.HeartbeatCheckInterval())
+
+	if qopts.IncludeUnhealthy() {
+		ticker.Stop()
+	}
 
-		service, err := getServiceFromValue(value, sid)
-		if err != nil {
-			s.opts.InstrumentsOptions().Logger().Errorf("could not unmarshal update from kv store for cluster")
-			errCounter.Inc(1)
-			continue
+	var err error
+
+	for {
+		select {
+		case <-vw.C():
+			value := vw.Get()
+			c.opts.InstrumentsOptions().Logger().Infof("received topology update notification on version %d", value.Version())
```

Comment:
nit: can do `log := c.opts.InstrumentsOptions().Logger()` outside the for loop so you can directly do `log.Infof(...)` and below